File: apps/opik-frontend/src/lib/dashboard/layout.ts
import {
  DashboardLayout,
  DashboardLayoutItem,
  DashboardWidget,
  WIDGET_TYPE,
  WidgetSize,
  WidgetSizeConfig,
} from "@/types/dashboard";
import isEmpty from "lodash/isEmpty";

export const GRID_COLUMNS = 6;
export const ROW_HEIGHT = 75;
export const GRID_MARGIN: [number, number] = [16, 16];
export const CONTAINER_PADDING: [number, number] = [0, 0];
export const MAX_WIDGET_HEIGHT = 12;
export const MIN_WIDGET_WIDTH = 1;
export const MIN_WIDGET_HEIGHT = 1;

export const getWidgetSizeConfig = (type: string): WidgetSizeConfig => {
  switch (type) {
    case WIDGET_TYPE.PROJECT_METRICS:
      return { w: 2, h: 4, minW: 2, minH: 4 };
    case WIDGET_TYPE.PROJECT_STATS_CARD:
      return { w: 1, h: 2, minW: 1, minH: 2 };
    case WIDGET_TYPE.TEXT_MARKDOWN:
      return { w: 2, h: 4, minW: 1, minH: 4 };
    case WIDGET_TYPE.EXPERIMENTS_FEEDBACK_SCORES:
      return { w: 2, h: 4, minW: 2, minH: 4 };
    case WIDGET_TYPE.EXPERIMENT_LEADERBOARD:
      return { w: 6, h: 6, minW: 4, minH: 4 };
    default:
      return { w: 2, h: 2, minW: MIN_WIDGET_WIDTH, minH: MIN_WIDGET_HEIGHT };
  }
};

export const getColumnHeights = (layout: DashboardLayout): number[] => {
  const heights = new Array(GRID_COLUMNS).fill(0);

  layout.forEach((item) => {
    const startCol = item.x;
    const endCol = Math.min(item.x + item.w, GRID_COLUMNS);
    const itemBottom = item.y + item.h;

    for (let col = startCol; col < endCol; col++) {
      heights[col] = Math.max(heights[col], itemBottom);
    }
  });

  return heights;
};

export const findFirstAvailablePosition = (
  w: number,
  h: number,
  columnHeights: number[],
): { x: number; y: number } => {
  let minHeight = Infinity;
  let bestX = 0;

  for (let x = 0; x <= GRID_COLUMNS - w; x++) {
    const maxHeightInRange = Math.max(...columnHeights.slice(x, x + w));

    if (maxHeightInRange < minHeight) {
      minHeight = maxHeightInRange;
      bestX = x;
    }
  }

  return { x: bestX, y: minHeight };
};

export const calculateLayoutForAddingWidget = (
  layout: DashboardLayout,
  widget: DashboardWidget,
  size?: WidgetSize,
): DashboardLayout => {
  const sizeConfig = getWidgetSizeConfig(widget.type);
  const { w, h } = size || { w: sizeConfig.w, h: sizeConfig.h };
  const { minW, minH } = sizeConfig;

  const newItem: DashboardLayoutItem = {
    i: widget.id,
    x: 0,
    y: 0,
    w,
    h,
    minW,
    minH,
    maxW: GRID_COLUMNS,
    maxH: MAX_WIDGET_HEIGHT,
  };

  if (isEmpty(layout)) {
    return [newItem];
  }

  const columnHeights = getColumnHeights(layout);
  const { x, y } = findFirstAvailablePosition(w, h, columnHeights);

  newItem.x = x;
  newItem.y = y;

  return [...layout, newItem];
};

export const normalizeLayout = (
  layout: DashboardLayout,
  widgets?: DashboardWidget[],
): DashboardLayout => {
  return layout.map((item) => {
    const widget = widgets?.find((w) => w.id === item.i);
    const { minW, minH } = widget
      ? getWidgetSizeConfig(widget.type)
      : { minW: MIN_WIDGET_WIDTH, minH: MIN_WIDGET_HEIGHT };

    return {
      i: item.i,
      x: Math.max(0, Math.min(item.x, GRID_COLUMNS - item.w)),
      y: Math.max(0, item.y),
      w: Math.max(minW, Math.min(item.w, GRID_COLUMNS)),
      h: Math.max(minH, Math.min(item.h, MAX_WIDGET_HEIGHT)),
      minW,
      minH,
      maxW: GRID_COLUMNS,
      maxH: MAX_WIDGET_HEIGHT,
    };
  });
};

export const removeWidgetFromLayout = (
  layout: DashboardLayout,
  widgetId: string,
): DashboardLayout => {
  return layout.filter((item) => item.i !== widgetId);
};

export const areLayoutsEqual = (
  prev: DashboardLayout,
  next: DashboardLayout,
): boolean => {
  if (prev.length !== next.length) return false;

  const sortedPrev = [...prev].sort((a, b) => a.i.localeCompare(b.i));
  const sortedNext = [...next].sort((a, b) => a.i.localeCompare(b.i));

  return sortedPrev.every((prevItem, i) => {
    const nextItem = sortedNext[i];
    return (
      prevItem.i === nextItem.i &&
      prevItem.x === nextItem.x &&
      prevItem.y === nextItem.y &&
      prevItem.w === nextItem.w &&
      prevItem.h === nextItem.h
    );
  });
};

export const getLayoutItemSize = (
  layout: DashboardLayout,
  widgetId: string,
): { w: number; h: number } | undefined => {
  const item = layout.find((item) => item.i === widgetId);
  return item ? { w: item.w, h: item.h } : undefined;
};

File: sdks/python/src/opik/api_objects/dashboard/layout.py
"""Grid auto-layout, ported from apps/opik-frontend/src/lib/dashboard/layout.ts.

Operates on :class:`~opik.api_objects.dashboard.types.DashboardLayoutItem` objects.
Keeping the algorithm in sync with the frontend means widgets the SDK adds are
positioned the same way the UI would position them.
"""

from typing import Any, Dict, List, Optional

from . import types
from .types import DashboardLayoutItem

GRID_COLUMNS = types.GRID_COLUMNS
MAX_WIDGET_HEIGHT = types.MAX_WIDGET_HEIGHT
MIN_WIDGET_WIDTH = types.MIN_WIDGET_WIDTH
MIN_WIDGET_HEIGHT = types.MIN_WIDGET_HEIGHT

_WIDGET_SIZE_CONFIG: Dict[str, Dict[str, int]] = {
    types.WidgetType.PROJECT_METRICS.value: {"w": 2, "h": 4, "minW": 2, "minH": 4},
    types.WidgetType.PROJECT_STATS_CARD.value: {"w": 1, "h": 2, "minW": 1, "minH": 2},
    types.WidgetType.TEXT_MARKDOWN.value: {"w": 2, "h": 4, "minW": 1, "minH": 4},
    types.WidgetType.EXPERIMENTS_FEEDBACK_SCORES.value: {
        "w": 2,
        "h": 4,
        "minW": 2,
        "minH": 4,
    },
    types.WidgetType.EXPERIMENT_LEADERBOARD.value: {
        "w": 6,
        "h": 6,
        "minW": 4,
        "minH": 4,
    },
}

_DEFAULT_SIZE_CONFIG = {
    "w": 2,
    "h": 2,
    "minW": MIN_WIDGET_WIDTH,
    "minH": MIN_WIDGET_HEIGHT,
}


def get_widget_size_config(widget_type: str) -> Dict[str, int]:
    return _WIDGET_SIZE_CONFIG.get(widget_type, _DEFAULT_SIZE_CONFIG)


def get_column_heights(layout: List[DashboardLayoutItem]) -> List[int]:
    heights = [0] * GRID_COLUMNS

    for item in layout:
        start_col = item.x
        end_col = min(item.x + item.w, GRID_COLUMNS)
        item_bottom = item.y + item.h

        for col in range(start_col, end_col):
            heights[col] = max(heights[col], item_bottom)

    return heights


def find_first_available_position(
    w: int, h: int, column_heights: List[int]
) -> Dict[str, int]:
    # The frontend leaves w unclamped, which makes the search range empty (and y
    # Infinity) for w > GRID_COLUMNS. Clamp here so the result is always a finite
    # placement; real widget sizes never exceed the grid width.
    w = min(w, GRID_COLUMNS)

    min_height = float("inf")
    best_x = 0

    for x in range(0, GRID_COLUMNS - w + 1):
        max_height_in_range = max(column_heights[x : x + w])
        if max_height_in_range < min_height:
            min_height = max_height_in_range
            best_x = x

    return {"x": best_x, "y": int(min_height)}


def calculate_layout_for_adding_widget(
    layout: List[DashboardLayoutItem],
    widget_type: str,
    widget_id: str,
    size: Optional[Dict[str, int]] = None,
) -> List[DashboardLayoutItem]:
    size_config = get_widget_size_config(widget_type)
    raw_w = size["w"] if size else size_config["w"]
    raw_h = size["h"] if size else size_config["h"]
    w = max(size_config["minW"], min(raw_w, GRID_COLUMNS))
    h = max(size_config["minH"], min(raw_h, MAX_WIDGET_HEIGHT))

    new_item = DashboardLayoutItem(
        id=widget_id,
        x=0,
        y=0,
        w=w,
        h=h,
        min_w=size_config["minW"],
        min_h=size_config["minH"],
        max_w=GRID_COLUMNS,
        max_h=MAX_WIDGET_HEIGHT,
    )

    if not layout:
        return [new_item]

    column_heights = get_column_heights(layout)
    position = find_first_available_position(w, h, column_heights)
    new_item.x = position["x"]
    new_item.y = position["y"]

    return [*layout, new_item]


def normalize_layout(
    layout: List[DashboardLayoutItem],
    widgets: Optional[List[Dict[str, Any]]] = None,
) -> List[DashboardLayoutItem]:
    widgets_by_id = {w["id"]: w for w in (widgets or [])}

    normalized: List[DashboardLayoutItem] = []
    for item in layout:
        widget = widgets_by_id.get(item.id)
        if widget is not None:
            size_config = get_widget_size_config(str(widget["type"]))
            min_w, min_h = size_config["minW"], size_config["minH"]
        else:
            min_w, min_h = MIN_WIDGET_WIDTH, MIN_WIDGET_HEIGHT

        normalized.append(
            DashboardLayoutItem(
                id=item.id,
                x=max(0, min(item.x, GRID_COLUMNS - item.w)),
                y=max(0, item.y),
                w=max(min_w, min(item.w, GRID_COLUMNS)),
                h=max(min_h, min(item.h, MAX_WIDGET_HEIGHT)),
                min_w=min_w,
                min_h=min_h,
                max_w=GRID_COLUMNS,
                max_h=MAX_WIDGET_HEIGHT,
            )
        )

    return normalized


def remove_widget_from_layout(
    layout: List[DashboardLayoutItem], widget_id: str
) -> List[DashboardLayoutItem]:
    return [item for item in layout if item.id != widget_id]
